### PR TITLE
Feature (crypto): BTC rainbow chart

### PR DESCRIPTION
**Pull request by jose-donato**
Adds new `btcrb` feature which displays bitcoin rainbow chart including halving dates. Inspired by the amazing [Blockchain Center website](https://www.blockchaincenter.net/)

`crypto/ov/btcrb`

image

Thanks @DidierRLopes for the help on making the chart 🙌

### CODE REVIEW

**Comment by piiq** [comment]
image

p.s. possible make the prices be in usd without the power?

**Comment by jose-donato** [comment]
> image
> 
> p.s. possible make the prices be in usd without the power?

Tried some stuff without success. I'm open to suggestions on how to do that 😄

**Comment by DidierRLopes** [comment]
See #1163 ! That should do, I tested with bitcoin and looked pretty :)

**Comment by DidierRLopes** [comment]
Could be fun to have our own "fit" to other coins to not work only on bitcoin.

I experimented with the code and we can try to do a scipy_fit on a similar equation and then adjust the weights according to the original formula. It's not as meaningful as bitcoin, but could be interesting :)

```
from scipy.optimize import curve_fit

d0 = datetime.strptime("2012-01-01", "%Y-%m-%d")
x = range((df_data.index[0]-d0).days, (df_data.index[-1]-d0).days+1)

def func(x, a, b):
    return 10 ** ((a * np.log(x+b)) - 19.463)

popt, pcov = curve_fit(func, x, df_data["Adj Close"].values, p0=[2.8445, 1293], maxfev=5000)

fig, ax = plt.subplots(figsize=(25,10))
d0 = datetime.strptime("2012-01-01", "%Y-%m-%d")
x = range((df_data.index[0]-d0).days, (df_data.index[-1]-d0).days+1)

y0 = func(x, (1400 * popt[0])/1293, (2.90 * popt[1])/2.8445)
y1 = func(x, (1375 * popt[0])/1293, (2.886 * popt[1])/2.8445)
y2 = func(x, (1350 * popt[0])/1293, (2.872 * popt[1])/2.8445)
y3 = func(x, (1320 * popt[0])/1293, (2.859 * popt[1])/2.8445)
y4 = func(x, popt[0], popt[1])
y5 = func(x, (1275 * popt[0])/1293, (2.8295 * popt[1])/2.8445)
y6 = func(x, (1250 * popt[0])/1293, (2.815 * popt[1])/2.8445)
y7 = func(x, (1225 * popt[0])/1293, (2.801 * popt[1])/2.8445)
y8 = func(x, (1200 * popt[0])/1293, (2.788 * popt[1])/2.8445)

....
```

Dropping here the code in case you want to use it. Looks like this:
ETHEREUM
image
LTC
image

**Comment by jmaslek** [comment]
The btc works because it is such an exponential growth.  If you wanted to extrapolate better you could look at a rolling variance and see if you can extrapolate a curve from that.

The non exponential fits seem are.....

**Comment by DidierRLopes** [comment]
> The btc works because it is such an exponential growth. If you wanted to extrapolate better you could look at a rolling variance and see if you can extrapolate a curve from that.
> 
> The non exponential fits seem are.....

Yea, the idea was to keep the exponential fit. If it doesn't happen, it's not as great :) No other coin will be as good as bitcoin. But you can understand their exponential growth from the rainbow behaviour, and that way we keep the formula.

**Comment by piiq** [comment]
@jose-donato the latest commit gives me a chart worth sharing
image

**Comment by jmaslek** [comment]
...can we just leave at bitcoin

**Comment by piiq** [comment]
> Tried some stuff without success. I'm open to suggestions on how to do that 😄

@jose-donato @DidierRLopes 
I've succeeded with a `FuncFormatter` to doing this
image

**Comment by jose-donato** [comment]
image

Some chart improvements. 

image
Do you think having the volume below would bring anything more or should we just keep it like it is right now?

**Comment by jose-donato** [comment]
> > Tried some stuff without success. I'm open to suggestions on how to do that 😄
> 
> @jose-donato @DidierRLopes I've succeeded with a `FuncFormatter` to doing this image

latest commit fixed some stuff but your formatter looks way better. will update it thanks!!

**Comment by DidierRLopes** [comment]
> stumbled into a breakpoint and gave me a chart worth sharing
> image

on that website they have a "uptrend" scenario which is pretty what you showed here, which is with a linear scale

**Comment by DidierRLopes** [comment]
I like the idea of adding the volume below! @jose-donato

**Comment by piiq** [comment]
> Do you think having the volume below would bring anything more

Only if the chart will remain zoomable after that. There is a chance zooming will be ugly if you add the volume to the same ax

**Comment by DidierRLopes** [comment]
What do you think about having this on the "ta"? Isn't it a bit of technical analysis stuff in the end? @jmaslek @piiq

**Comment by piiq** [comment]
> What do you think about having this on the "ta"? Isn't it a bit of technical analysis stuff in the end? @jmaslek @piiq

If rearranging stuff I would only ask to put the corresponding view and model into the relevant context folder. Now the controller is dd but the view and the model are in overview.
Regarding where it should be - you decide. This is a "meme" feature so I wouldn't hide it too deep

**Comment by jmaslek** [comment]
> > What do you think about having this on the "ta"? Isn't it a bit of technical analysis stuff in the end? @jmaslek @piiq
> 
> If rearranging stuff I would only ask to put the corresponding view and model into the relevant context folder. Now the controller is dd but the view and the model are in overview. Regarding where it should be - you decide. This is a "meme" feature so I wouldn't hide it too deep

Not ta

**Comment by DidierRLopes** [comment]
> > > What do you think about having this on the "ta"? Isn't it a bit of technical analysis stuff in the end? @jmaslek @piiq
> > 
> > 
> > If rearranging stuff I would only ask to put the corresponding view and model into the relevant context folder. Now the controller is dd but the view and the model are in overview. Regarding where it should be - you decide. This is a "meme" feature so I wouldn't hide it too deep
> 
> Not ta

I suggested "ta" because of being kinda of an "RSI" levels with the "oversold/overbought"

**Review by jmaslek — APPROVED** [review]

File: gamestonk_terminal/cryptocurrency/due_diligence/glassnode_model.py [hunk 1]
```diff
@@ -163,6 +163,47 @@
 INTERVALS = ["1h", "24h", "10m", "1w", "1month"]
 
 
+def get_close_price(asset: str, interval: str, since: int, until: int) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (jose-donato) [reply gamestonk_terminal/cryptocurrency/due_diligence/glassnode_model.py:170]: Na, this is one of the few sources that I found that go until BTC inception. It does not provide volume though :(

**Pull request merged**